azuredevops/internal/service/taskagent/resource_environment.go

```diff
@@ -0,0 +1,199 @@
+package taskagent
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/google/uuid"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+	"github.com/microsoft/azure-devops-go-api/azuredevops/v6/taskagent"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/client"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/converter"
+	"github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/utils/tfhelper"
+)
+
+const (
+	envProjectId   = "project_id"
+	envName        = "name"
+	envDescription = "description"
+)
+
+// ResourceEnvironment schema and implementation for environment resource
+func ResourceEnvironment() *schema.Resource {
+	return &schema.Resource{
+		Create:   resourceEnvironmentCreate,
+		Read:     resourceEnvironmentRead,
+		Update:   resourceEnvironmentUpdate,
+		Delete:   resourceEnvironmentDelete,
+		Importer: tfhelper.ImportProjectQualifiedResourceInteger(),
+		Schema: map[string]*schema.Schema{
+			envProjectId: {
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IsUUID,
+			},
+			envName: {
```

Comment:
oh man, the environment name has a strange behavior?
when you do 127 characters the error message changes and it shows it would like to append the project-id to the environment name. It is not what it actually expects: it wants to append the projects name, or rather the combination of environment name and project name can't exceed the 127 characters for it to be valid it seems